src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java

```diff
@@ -106,72 +266,186 @@ public void run() {
     }
 
+    private void updateOnOff(OnOffType onoff) {
+        updateBrightness(lastBrightness);
+    }
+
+    private void updateBrightness(PercentType brightness) {
+        HSBType oldHSB = currentHSB;
+        HSBType newHSB = new HSBType(oldHSB.getHue(), oldHSB.getSaturation(), brightness);
+        currentHSB = newHSB;
+        lastBrightness = brightness;
+        updateChannelState(newHSB);
+    }
+
+    private void updateColorHSB(DecimalType hue, PercentType saturation) {
+        HSBType oldHSB = currentHSB;
+        HSBType newHSB = new HSBType(hue, saturation, oldHSB.getBrightness());
+        currentHSB = newHSB;
+        updateChannelState(newHSB);
+    }
+
+    private void updateColorXY(PercentType x, PercentType y) {
+        logger.debug("{}: Update Color XY. {}, {}", endpoint.getIeeeAddress(), x.toString(), y.toString());
+        HSBType color = ColorHelper.fromXY(x.floatValue() / 100.0f, y.floatValue() / 100.0f, 1.0f);
+        logger.debug("{}: Calculated Hue/Saturation. {}, {}", endpoint.getIeeeAddress(), color.getHue(),
+                color.getSaturation());
+        updateColorHSB(color.getHue(), color.getSaturation());
+    }
+
+    private void updateColorHSB() {
+        DecimalType hue = new DecimalType(Float.valueOf(lastHue).toString());
+        PercentType saturation = new PercentType(Float.valueOf(lastSaturation).toString());
+        updateColorHSB(hue, saturation);
+        hueChanged = false;
+        saturationChanged = false;
+    }
+
+    private void updateColorXY() {
+        PercentType x = new PercentType(Float.valueOf(lastX * 100.0f).toString());
+        PercentType y = new PercentType(Float.valueOf(lastY * 100.0f).toString());
+        updateColorXY(x, y);
+        xChanged = false;
+        yChanged = false;
+    }
+
     @Override
     public void attributeUpdated(ZclAttribute attribute) {
         logger.debug("ZigBee attribute reports {} from {}", attribute, endpoint.getIeeeAddress());
-        if (attribute.getCluster() == ZclClusterType.COLOR_CONTROL) {
-            if (attribute.getId() == ZclColorControlCluster.ATTR_CURRENTHUE) {
-                Integer value = (Integer) attribute.getLastValue();
-                if (value != null) {
-                    DecimalType hue = new DecimalType(value * 100 / 254);
-                    currentHSB = new HSBType(hue, currentHSB.getSaturation(), currentHSB.getBrightness());
-                    updateChannelState(currentHSB);
+
+        synchronized (colorUpdateSync) {
+            try {
+                if (attribute.getCluster().getId() == ZclOnOffCluster.CLUSTER_ID
+                        && attribute.getId() == ZclOnOffCluster.ATTR_ONOFF) {
+                    Boolean value = (Boolean) attribute.getLastValue();
+                    OnOffType onoff = value ? OnOffType.ON : OnOffType.OFF;
+                    updateOnOff(onoff);
+                } else if (attribute.getCluster().getId() == ZclLevelControlCluster.CLUSTER_ID
```

Comment:
Will do. I do not normally like nested ifs for code readability if the repeated condition is simple to compute, but I understand it will be more efficient anyway.